### PR TITLE
devicetree: improve output

### PR DESCRIPTION
**Pull request by mbolivar**
This makes the results a little easier to read, since node output appears in the same order as the top level comment summarizing the tree. It also makes gaps where nodes don't have matching compats more obvious.

### CODE REVIEW

**Review by ulfalizer — APPROVED** [review]
No strong opinions.

Maybe the comment at the top of the generated header could mention how the nodes are sorted.

**Review by pabigot — APPROVED** [review]
I agree this order is nicer to deal with.  It's not clear what the default order is, possibly depth-first or breadth-first.

I looked a little at possibly sorting the `node` property in this way, so the order would be consistent for all iterations, but that destabilizes the order of flash partitions.  So IMO it's best to leave this sort for this output.

**Comment by mbolivar** [comment]
> Maybe the comment at the top of the generated header could mention how the nodes are sorted.

I didn't want to do that because it wasn't totally true: the special casing around flash partitions is an exception, and I didn't want to take the time to fix it.

But you're right that that would be better, so it's fixed now. I'm also emitting comments for all nodes which should make it clearer why output for one might be missing (e.g. if there is no matching compatible or the node is disabled).

I tacked on what I consider a fix for DT_COMPAT_FIXED_PARTITIONS. I can split that into a separate PR if anyone wants.

**Comment by mbolivar** [comment]
> I looked a little at possibly sorting the `node` property in this way, so the order would be consistent for all iterations, but that destabilizes the order of flash partitions. So IMO it's best to leave this sort for this output.

@pabigot note that the latest change has removed the property that flash partitions are sorted by start address, but I think it's better for consistency to keep it this way actually.

Can we add a tiebreaker to scc_order that considers the unit address when assigning ordinals to sibling nodes in the tree somehow? I haven't looked at the implementation.

**Comment by mbolivar** [comment]
> Can we add a tiebreaker to scc_order that considers the unit address when assigning ordinals to sibling nodes in the tree somehow? I haven't looked at the implementation.

Added a commit that does just that.

**Comment by pabigot** [comment]
This still looks fine, modulo the need to rebase.  Thanks for making it more consistent.

**Comment by mbolivar** [comment]
Rebased, thanks.

File: scripts/dts/gen_defines.py [hunk 1]
```diff
@@ -67,11 +72,10 @@ def main():
         #define DT_COMPAT_<COMPAT> 1
         out(f"COMPAT_{str2ident(compat)}", 1)
 
-    # Derived from /chosen
+    # Definitions derived from /chosen nodes.
```
Comment (ulfalizer) [inline scripts/dts/gen_defines.py]: I skip the period when it's just one sentence, so there's a method to it at least.

File: scripts/dts/gen_defines.py [hunk 2]
```diff
@@ -366,20 +386,15 @@ def write_addr_size(edt, prop_name, prefix):
 
 
 def write_flash(edt):
-    # Writes flash-related output
+    # Writes chosen and tree-wide flash-related output
+    global flash_area_num
```
Comment (ulfalizer) [inline scripts/dts/gen_defines.py]: I only put in `global` if it's written to, but fine if it was meant as a hint.

File: scripts/dts/gen_defines.py [hunk 3]
```diff
@@ -792,7 +805,8 @@ def out_comment(s, blank_before=True):
         #   /* foo bar */
         print("/* " + s + " */", file=header_file)
 
-    print("\n".join("# " + line for line in s.splitlines()), file=conf_file)
+    print("\n".join("#" + (' ' + line if line.strip() else '')
```
Comment (ulfalizer) [inline scripts/dts/gen_defines.py]: Maybe

```python
"# " + line if line.strip() else "#"
```

**Comment by mbolivar** [comment]
@ulfalizer comments addressed, thanks!

**Pull request merged**